swarm/network/stream/stream.go

```diff
@@ -0,0 +1,441 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package stream
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+const (
+	Low uint8 = iota
+	Mid
+	High
+	Top
+	PriorityQueue         // number of queues
+	PriorityQueueCap = 32 // queue capacity
+	HashSize         = 32
+)
+
+// Registry registry for outgoing and incoming streamer constructors
+type Registry struct {
+	api         *API
+	addr        *network.BzzAddr
+	skipCheck   bool
+	clientMu    sync.RWMutex
+	serverMu    sync.RWMutex
+	peersMu     sync.RWMutex
+	serverFuncs map[string]func(*Peer, []byte) (Server, error)
+	clientFuncs map[string]func(*Peer, []byte) (Client, error)
+	peers       map[discover.NodeID]*Peer
+	delivery    *Delivery
+	store       storage.ChunkStore
+}
+
+// NewRegistry is Streamer constructor
+func NewRegistry(addr *network.BzzAddr, delivery *Delivery, store storage.ChunkStore, skipCheck bool) *Registry {
+	streamer := &Registry{
+		addr:        addr,
+		skipCheck:   skipCheck,
+		store:       store,
+		serverFuncs: make(map[string]func(*Peer, []byte) (Server, error)),
+		clientFuncs: make(map[string]func(*Peer, []byte) (Client, error)),
+		peers:       make(map[discover.NodeID]*Peer),
+		delivery:    delivery,
+	}
+	streamer.api = NewAPI(streamer, streamer.store)
+	delivery.getPeer = streamer.getPeer
+	streamer.RegisterServerFunc(swarmChunkServerStreamName, func(_ *Peer, t []byte) (Server, error) {
+		return NewSwarmChunkServer(delivery.db), nil
+	})
+	streamer.RegisterClientFunc(swarmChunkServerStreamName, func(p *Peer, t []byte) (Client, error) {
+		return NewSwarmSyncerClient(p, delivery.db, nil)
+	})
```

Comment:
I find it confusing that `NewRegistry` registers a syncer client and a chunk server, but not a syncer server? Is there a reason for this?

It's a little hard to follow how `Registry` / `Delivery` and the various `Client` / `Server` implementations all fit together, it would be good to have some high level documentation.